RetroBar/Utilities/ClockFlyoutLauncher.cs

```diff
@@ -0,0 +1,147 @@
+﻿using ManagedShell.AppBar;
+using ManagedShell.Common.Helpers;
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Windows;
+using static ManagedShell.Interop.NativeMethods;
+using Rect = ManagedShell.Interop.NativeMethods.Rect;
+using Screen = System.Windows.Forms.Screen;
+
+namespace RetroBar.Utilities
+{
+    internal class ClockFlyoutLauncher
+    {
+        private const SetWindowPosFlags NoPosFlags = SetWindowPosFlags.SWP_NOSIZE | SetWindowPosFlags.SWP_NOACTIVATE | SetWindowPosFlags.SWP_NOZORDER;
+
+        #region Aero clock
+
+        private static readonly Lazy<Type> AeroClockType = new(() => Type.GetTypeFromCLSID(new("A323554A-0FE1-4E49-AEE1-6722465D799F")));
+        private static object _aeroClockInstance;
+
+        [ComImport, Guid("7A5FCA8A-76B1-44C8-A97C-E7173CCA5F4F"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+        private interface IAeroClock // For 8+
+        {
+            [PreserveSig]
+            void ShowFlyout(IntPtr hWnd, ref Rect lpRect);
+        }
+
+        [ComImport, Guid("4376DF10-A662-420B-B30D-958881461EF9"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+        private interface IAeroClockLegacy // For Vista and 7
+        {
+            [PreserveSig]
+            void ShowFlyout(int unk, ref Rect lpRect);
+        }
+
+        internal static void ShowAeroClockFlyout(IntPtr taskbarHwnd)
+        {
+            Screen taskbarScreen = Screen.FromHandle(taskbarHwnd);
+            if (!GetWindowRect(taskbarHwnd, out Rect lpRect))
+            {
+                return;
+            }
+
+            _aeroClockInstance ??= Activator.CreateInstance(AeroClockType.Value);
+
+            if (ManagedShell.Common.Helpers.EnvironmentHelper.IsWindows8OrBetter)
+            {
+                ((IAeroClock)_aeroClockInstance).ShowFlyout(taskbarHwnd, ref lpRect);
+            }
+            else
+            {
+                ((IAeroClockLegacy)_aeroClockInstance).ShowFlyout(0, ref lpRect);
+            }
+
+            FixAeroClockFlyoutPosition(taskbarScreen);
+        }
+
+        private static void FixAeroClockFlyoutPosition(Screen taskbarScreen)
+        {
+            IntPtr clockFlyoutHwnd = FindWindow("ClockFlyoutWindow", null);
+            if (clockFlyoutHwnd == IntPtr.Zero)
+            {
+                return;
+            }
+
+            GetWindowRect(clockFlyoutHwnd, out Rect clockFlyoutRect);
+
+            // Keep the flyout at least 3 by 3 pixels inside the working area
+            int newX = (int)Math.Max(taskbarScreen.WorkingArea.Left + 3, Math.Min(taskbarScreen.WorkingArea.Right - (clockFlyoutRect.Width) - 3, clockFlyoutRect.Left));
+            int newY = (int)Math.Max(taskbarScreen.WorkingArea.Top + 3, Math.Min(taskbarScreen.WorkingArea.Bottom - (clockFlyoutRect.Height) - 3, clockFlyoutRect.Top));
+            SetWindowPos(clockFlyoutHwnd, IntPtr.Zero, newX, newY, 0, 0, (int)(NoPosFlags));
```

Comment:
Why 3?

Reply:
The last time I checked, it was the margin (spacing?) that separated the taskbar from the clock flyout in Windows 7, though I haven't checked if it applies to other screens with different scaling.

Reply:
Hmm, I just compared and my Win7 has 2px additional on the X and 1px additional on the Y, at least with Aero.

Reply:
Just to found out that Vista does not use spacing (no floating flyout)...
![VistaClockFlyoutNoSpacing](https://github.com/user-attachments/assets/33e234d1-ede2-410d-9bf0-ad652ccfaf84)

Maybe this should be customizable by the theme.

Reply:
Given that RetroBar targets emulating Vista and below, let's remove the margin for now.